rust/src/composepost.rs

```diff
@@ -202,6 +203,44 @@ fn postprocess_presets(rootfs_dfd: &Dir) -> Result<()> {
     Ok(())
 }
 
+fn is_overlay_whiteout(meta: &cap_std::fs::Metadata) -> bool {
+    (meta.mode() & libc::S_IFCHR) > 0 && meta.rdev() == 0
+}
+
+/// Auto-synthesize embedded overlayfs whiteouts; for more information
+/// see https://github.com/ostreedev/ostree/pull/2722/commits/0085494e350c72599fc5c0e00422885d80b3c660
+#[context("Postprocessing embedded overlayfs")]
+fn postprocess_embedded_ovl_whiteouts(root: &Dir) -> Result<()> {
+    const OSTREE_WHITEOUT_PREFIX: &str = ".ostree-wh.";
+    fn recurse(root: &Dir, path: &Path) -> Result<u32> {
+        let mut n = 0;
+        for entry in root.read_dir(path)? {
+            let entry = entry?;
+            let meta = entry.metadata()?;
+            let name = entry.file_name();
+            if meta.is_dir() {
+                n += recurse(root, &path.join(name))?;
+                continue;
+            }
+            if !is_overlay_whiteout(&meta) {
+                continue;
+            };
+            let srcpath = path.join(&name);
+            let mut targetname = OsStr::new(OSTREE_WHITEOUT_PREFIX).to_owned();
+            targetname.push(&name);
+            let destpath = path.join(&targetname);
+            root.rename(&srcpath, root, &destpath)
+                .with_context(|| format!("Renaming overlay whiteout: {srcpath:?}"))?;
```

Comment:
I thought the problem was that `ostree` cannot handle character devices on commit; I assumed you had to re-create the whiteout as regular file? From my simple `org.osbuild.ostree.whiteout` stage I wrote that aimed at doing the same thing:
```
def main(tree):
    for root, _, files in os.walk(tree, topdown=False):
        for name in files:
            path = os.path.join(root, name)

            statinfo = os.stat(path)
            if not is_whiteout(statinfo):
                continue

            whiteout = os.path.join(root, f".ostree-wh.{name}")
            os.truncate(whiteout, 0)
            shutil.copystat(path, whiteout)
            os.unlink(path)
```
I might be missing something.

Reply:
Oh no, the code was just broken (in multiple ways), and I was confused by the error I was seeing and wanted to see if CI hit the same problem I did locally, and it did.

Updated and should be good now.